### PR TITLE
Use ReshapedArray for isbitsunion arrays (fixes #28611)

### LINKED ISSUE #28611: Regression: reshape() after push!() with missing data returns wrong result
On 0.7:

```
julia> v = allowmissing(Float64[])
0-element Array{Union{Missing, Float64},1}

julia> for i in 1:10 push!(v, i) end

julia> v
10-element Array{Union{Missing, Float64},1}:
  1.0
  2.0
  3.0
  4.0
  5.0
  6.0
  7.0
  8.0
  9.0
 10.0

julia> v[5] = missing
missing

julia> reshape(v, 2,5)
2×5 Array{Union{Missing, Float64},2}:
 1.0       missing  missing  missing  missing
  missing  missing  missing  missing  missing
```

### PR DESCRIPTION
**Pull request by timholy**
The only non-obvious part of this is getting correct dispatch for `Array{Any}`, ensuring that we return an `Array` without messing up inference of other cases. (It fails https://github.com/JuliaLang/julia/blob/f4c777972e58d25d4636480d8e5ca30066a6235e/test/core.jl#L4172-L4174 if we don't do this properly.)

Fixes #28611.

CC @GlenHertz, @nalimilan.

### CODE REVIEW

**Comment by timholy** [comment]
The failure on apple seems spurious.

**Comment by nalimilan** [comment]
Thanks! But don't we need a more general fix handling all unions of `isbits` types?

**Comment by timholy** [comment]
Hmm, good point. I had forgotten this was no longer special-cased for Missing.

**Comment by timholy** [comment]
Technically this is breaking, although the cases it breaks are already silently buggy. If the breaking nature is a problem, perhaps the backport label should be removed.

File: base/array.jl [hunk 1]
```diff
@@ -172,7 +172,7 @@ julia> Base.isbitsunion(Union{Float64, String})
 false
 ```
 """
-isbitsunion(u::Union) = ccall(:jl_array_store_unboxed, Cint, (Any,), u) != Cint(0)
+isbitsunion(u::Union) = (@_pure_meta; ccall(:jl_array_store_unboxed, Cint, (Any,), u) != Cint(0))
```
Comment (timholy) [inline base/array.jl]: @vtjnash, is this OK?

Comment (vtjnash) [reply base/array.jl]: Yes (we mark many of these methods pure for this reason)

**Comment by timholy** [comment]
I'll merge soon, barring further concerns.

**Comment by KristofferC** [comment]
This doesn't seem backportable.

**Comment by andreasnoack** [comment]
What about the bug in #28611

**Comment by KristofferC** [comment]
It's in the same spot as those reffed by this PR: https://github.com/JuliaLang/julia/pull/29406.

**Comment by timholy** [comment]
> What about the bug in #28611

That's what this fixes (see the test). I'm not sure I understand the relevance of #29406.

**Comment by KristofferC** [comment]
I think Andreas point is that this should be backported because it fixes a bug. My response is that even though it fixes a bug it might break working code (which is the same as the bugs getting fixed by #29406). AFAIU, we are quite conservative when it comes to behavior changes for patch releases.

**Pull request merged**